### DISCUSSION TITLE
Add support for Mastodon

### DISCUSSION THREAD

**Discussion by CIAvash** [lowlighter/metrics#318] [Ideas]
Add support for showing latest toots from Mastodon.

**Comment by lowlighter** [comment]
I don't know mastodon enough to implement this, but you can submit a pull request to integrate this if you want 🙂

You just need to handle the api calls to return title, desc, cover image, etc. there:
https://github.com/lowlighter/metrics/blob/38e85eec1182480d2ea3c2a37ceff2d5c63cdeeb/source/plugins/posts/index.mjs#L16-L35

And add `mastodon` as a valid source in the `metadata.yml`:
https://github.com/lowlighter/metrics/blob/38e85eec1182480d2ea3c2a37ceff2d5c63cdeeb/source/plugins/posts/metadata.yml#L21-L23

Along with a mocked test in [`tests.yml`](https://github.com/lowlighter/metrics/blob/master/source/plugins/posts/tests.yml)

**Reply by Andre601** [reply]
From what I gather would you be able to use [`https://<domaind>.<tld>/api/v1/accounts/<account_id>/statuses`](https://docs.joinmastodon.org/client/public/#posts) to receive a list of [Status](https://docs.joinmastodon.org/entities/Status/) objects in JSON form.

The issue I see here is, that there isn't just one Mastodon page similar to what dev.to and hashnode is.
It would require at the very least these additional options:

- Domain used for the instance to fetch data from
- User ID to fetch data from
- Auth token (Maybe? I assume since we access public data that authorization is not required (Unless to avoid maybe API rate limits?), but perhaps only partial data will then be returned out of privacy...)

So, not entirely sure if a system similar to the post plugin's way would work here unless we can add additional options for it to check...
But personally would I think it's better to consider having a separate plugin than adding options only used for one specific site/service.

**Comment by Andre601** [comment]
Okay... I *tried* (keyword) to get a rough idea of a mastodon plugin...

https://github.com/Andre601/metrics/commit/0af780fbe56d82b1fa07f7011ad0e2220401c8e4

I have absolutely no idea if that would be the right aproach here to do, as I do not know JavaScript and honestly just used the posts plugin as a template.

The reason for it being a separate plugin instead of being included within mastodon was stated in my above comment, namely that there isn't a single instance, therefore requiring the user to provide a domain, which as a separate posts option feels kinda like bloat to have here.

Also, I want to point out one smaller issue with the current setup, which is that it requires a User ID, not name, for the API to return data. And user IDs aren't as easy to get for a normal user.
So perhaps it can be considered to add an extra step that manages the `username -> user id` lookup. This is as simple as calling the `/api/v1/accounts/lookup` API path of a Mastodon instance and provide a `acct` query parameter with the username. This should return a JSON containing the user ID.

I would implement that... But as mentioned no JavaScript knowledge so can't really work it out.

If anything is this meant as a sort of "proof of concept" to maybe even give ideas to work with...

And on a kinda related note: Not sure if anything else (i.e. backend) would need an update too to work with this stuff...
The biggest concern for me is the fact that instances can have custom emojis which users can have in their posts and usernames, so this is also something to keep in mind...

**Reply by Andre601** [reply]
Small update: Made a ejs file for the plugin... Again using the posts plugin as inspiration.

Only thing I kinda fear is, that the SVG I used (The very first) is not properly resized... It's viewbox is 24x24 since it's from Simple Icons).

https://github.com/Andre601/metrics/commit/b1a6dc0566e39de5aa8d0f40b4527148dbf98aec
